indexing-service/src/main/java/org/apache/druid/indexing/overlord/ThreadingTaskRunner.java

```diff
@@ -0,0 +1,566 @@
+    extends BaseRestorableTaskRunner<ThreadingTaskRunner.ThreadingTaskRunnerWorkItem>
+    implements TaskLogStreamer, QuerySegmentWalker
+{
+  private static final EmittingLogger LOGGER = new EmittingLogger(ThreadingTaskRunner.class);
+
+  private final TaskToolboxFactory toolboxFactory;
+  private final TaskLogPusher taskLogPusher;
+  private final DruidNode node;
+  private final AppenderatorsManager appenderatorsManager;
+  private final TaskReportFileWriter taskReportFileWriter;
+  private final ListeningExecutorService taskExecutor;
+  private final ListeningExecutorService controlThreadExecutor;
+
+  private volatile boolean stopping = false;
+
+  @Inject
+  public ThreadingTaskRunner(
+      TaskToolboxFactory toolboxFactory,
+      TaskConfig taskConfig,
+      WorkerConfig workerConfig,
+      TaskLogPusher taskLogPusher,
+      ObjectMapper jsonMapper,
+      AppenderatorsManager appenderatorsManager,
+      TaskReportFileWriter taskReportFileWriter,
+      @Self DruidNode node
+  )
+  {
+    super(jsonMapper, taskConfig);
+    this.toolboxFactory = toolboxFactory;
+    this.taskLogPusher = taskLogPusher;
+    this.node = node;
+    this.appenderatorsManager = appenderatorsManager;
+    this.taskReportFileWriter = taskReportFileWriter;
+    this.taskExecutor = MoreExecutors.listeningDecorator(
+        Execs.multiThreaded(workerConfig.getCapacity(), "threading-task-runner-executor-%d")
+    );
+    this.controlThreadExecutor = MoreExecutors.listeningDecorator(
+        Execs.multiThreaded(workerConfig.getCapacity() * 2, "threading-task-runner-control-%d")
+    );
+  }
+
+  @Override
+  public Optional<ByteSource> streamTaskLog(String taskid, long offset) throws IOException
+  {
+    // task logs will appear in the main indexer log, streaming individual task logs is not supported
+    return Optional.absent();
+  }
+
+  @Override
+  public void start()
+  {
+    // Nothing to start.
+  }
+
+  @Override
+  public ListenableFuture<TaskStatus> run(Task task)
+  {
+    synchronized (tasks) {
+      tasks.computeIfAbsent(
+          task.getId(), k ->
+              new ThreadingTaskRunnerWorkItem(
+                  task,
+                  controlThreadExecutor.submit(
+                      new Callable<TaskStatus>() {
+                        @Override
+                        public TaskStatus call()
+                        {
+                          final String attemptUUID = UUID.randomUUID().toString();
+                          final File taskDir = taskConfig.getTaskDir(task.getId());
+                          final File attemptDir = new File(taskDir, attemptUUID);
+
+                          final TaskLocation taskLocation = TaskLocation.create(
+                              node.getHost(),
+                              node.getPlaintextPort(),
+                              node.getTlsPort()
+                          );
+
+                          try {
+                            final Closer closer = Closer.create();
```

Comment:
couldn't understand why we have this Closer object, nothing appears to be accumulated in it for closing.

Reply:
Thanks, removed this, it was a relic from ForkingTaskRunner